pymagicc/input.py

```diff
@@ -0,0 +1,225 @@
+        units = stream.readline().split()[1:]
+        regions = stream.readline().split()[1:]
+        index = pd.MultiIndex.from_arrays([gases, regions], names=['GAS', 'REGION'])
+        df = pd.read_csv(
+            stream,
+            skip_blank_lines=True,
+            delim_whitespace=True,
+            names=None,
+            header=None,
+            index_col=0)
+        df.index.name = 'YEAR'
+        df.columns = index
+
+        return df, self._extract_units(gases, units)
+
+    def _extract_units(self, gases, units):
+        combos = set(zip(gases, units))
+        result = {}
+        for v, u in combos:
+            if v not in result:
+                result[v] = u
+            else:
+                # this isn't expected to happen, but should check anyway
+                raise ValueError('Different units for {} in {}'.format(v, self.fname))
+
+        return result
+
+
+_file_types = {
+    'MAGICC6': MAGICC6Reader,
+    'MAGICC7': MAGICC7Reader,
+}
+
+
+def get_reader(fname):
+    with open(fname) as f:
+        lines = f.readlines()
+
+    # Infer the file type from the header
+    if '.__  __          _____ _____ _____ _____   ______   ______ __  __ _____  _____  _____ _   _' \
+            in lines[0]:
+        file_type = 'MAGICC7'
+    else:
+        file_type = 'MAGICC6'
+
+    return _file_types[file_type](fname, lines)
+
+
+class MAGICCInput(object):
+    """
+    An interface to read (and in future write) the input files used by MAGICC.
+
+    MAGICCInput can read input files from both MAGICC6 and MAGICC7. These
+    include files with extensions .IN and .SCEN7.
+
+    The MAGICCInput, once the target input file has been loaded it can be
+     treated as a pandas DataFrame. All the methods available to a DataFrame
+     can be called on the MAGICCInput.
+
+
+    >>> with MAGICC6() as magicc:
+    >>>     mdata = MAGICCInput('HISTRCP_CO2I_EMIS.IN')
+    >>>     mdata.read(magicc.run_dir)
+    >>>     mdata.plot()
+    """
+
+    def __init__(self, file_name=None):
+        """
+        Initialise a Input file object.
+
+        Optionally you can specify the filename of the target file. The file is
+        not read until the search directory is provided in `read`. This allows
+        for MAGICCInput files to be lazy-loaded once the appropriate MAGICC run
+        directory is known.
+        :param file_name: Optional file name, including extension for the target
+         file, i.e. 'HISTRCP_CO2I_EMIS.IN'
+        """
+        self.df = None
+        self.metadata = {}
+        self.name = file_name
+
+    def __getitem__(self, item):
+        """
+        Allow for indexing like a pandas DataFrame
+
+        >>> inpt = MAGICCInput('HISTRCP_CO2_CONC.IN')
+        >>> inpt.read('./')
+        >>> assert (inpt['CO2']['GLOBAL'] == inpt.df['CO2']['GLOBAL']).all()
+        """
+        if not self.is_loaded:
+            raise ValueError('File has not been read from disk yet')
+        return self.df[item]
```

Comment:
super nice, this could be where we could keep units as a column but not need to specify it when getting it?